Robbie Hilado - Front End Challenge

sopasdigital opened the pull request:
This is my attempt at doing the front end challenge for Signfica.

I was able to create the core functionality of the app except for 'favourite' functionality. Given the time limit, I was not able to apply all the proper styling presented in the style guide. I'm also aware that I'm missing some core states for the components like when there's no input from the user as of yet, if there are no poster images available to display, etc.

Danilo was a big help as he gave me some tips and suggestions on how to go about sending necessary state props to display information on the movie details page.

pbrandone commented:
Thanks for the submission Robbie 🎉

pdrbrnd commented on my-code/movie-search/src/Actions/index.js:
```diff
@@ -0,0 +1,25 @@
+const API_KEY = '5766cd6e';
+
+export function fetchMovies(searchTerm) {
+  const request = axios.get(`${ROOT_URL}${API_KEY}&s=${searchTerm}`);
```
given that `axios` returns a Promise, you're potentially giving your reducer a pending Promise payload.

The goal for these kind of features is to make sure you can take care of asynchronous calls (like an API request). Async calls have two distinct moments:
1. When you start to make the call
2. When there's an answer (or timeout).

It's always a good practice to handle correctly and create appropriate UI for these two moments by, for instance, creating an `isFetching` flag in your reducers or component state to show a loader when needed.
Also, we need to catch any kind of information if the request fails for some reason.

A simple solution (albeit very naïve, untested and whatnot) would be something like:
```jsx
class YourComponent extends Component {
  state = { 
    searchTerm: '',
    isFetching: false,
    results: [],
    error: false,
  }

  fetchMovies = () => {
    this.setState({ isFetching: true })

    axios.get(`${ROOT_URL}${API_KEY}&s=${searchTerm}`)
    .then(response => this.setState({ results: response.data, isFetching: false }))
    .catch(() => this.setState({ error: true, isFetching: false }))
  }

  // ...
}
```

If Redux is your thing, `redux-thunk` is a simple middleware that allows you to return functions for your action creators, thus making possible the delay of the dispatch.

```jsx
export const fetchMovies = searchTerm => dispatch => {
  dispatch({ type: 'FETCH_MOVIES_START' });

  axios.get(`${ROOT_URL}${API_KEY}&s=${searchTerm}`)
  .then(response => dispatch({ type: 'FETCH_MOVIES_COMPLETE', payload: response.data }))
  .catch(error => dispatch({ type: 'FETCH_MOVIES_ERROR', payload: error, error: true }))
}
```
or something like this.
This way we have 3 distinct action types (start, complete, error) and can make the UI react to each situation

(All this code was written here in the comments and is naive and untested)

sopasdigital replied:
This is great information, Pedro! I was wondering what would be good way of catching these eventualities. I'd love to add this to what I currently have.

I have yet to try `redux-thunk`, but I think it would be a cool inclusion. I still need to grasp the idea of states, I guess.

Would it be better I learn the 'simple solution' method first so I understand what's happening in vanillaJS before using `redux-thunk`?

pdrbrnd commented on my-code/movie-search/src/App.js:
```diff
@@ -0,0 +1,33 @@
+import promise from 'redux-promise';
+import { BrowserRouter, Route, Switch } from 'react-router-dom';
+
+import reducers from './Reducers';
```
The Capitalized words have a semantic meaning and I would avoid naming folders like "Actions", "Reducers", "Components" or "Containers".

It's best to leave capitalisation to components only

sopasdigital replied:
Ah! Okay, duly noted!

pdrbrnd commented on my-code/movie-search/src/App.js:
```diff
@@ -0,0 +1,33 @@
+
+const createStoreWithMiddleware = applyMiddleware(promise)(createStore);
+
+class App extends Component {
```
this can be refactored as a functional component

pdrbrnd commented on my-code/movie-search/src/Assets/icons/icon-arrow-grey.svg:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
A bit off-topic but it's always a good idea to compress svgs using something like SVGOMG

sopasdigital replied:
Ohh! I'll look into that, thanks for the suggestion!

pdrbrnd commented on my-code/movie-search/src/Components/AppLogo/index.js:
```diff
@@ -0,0 +1,5 @@
+import React from 'react';
+
+const AppLogo = (props) => <div className="AppLogo"><img src={require('../../Assets/logos/logo.svg')} alt="What's in"/></div>
```
Why this `require`?

You could
```jsx
import React from 'react';
import logo from '../../Assets/logos/logo.svg';
```

sopasdigital replied:
You're right, I should've done that instead.

pdrbrnd commented on my-code/movie-search/src/Containers/MovieCatalogue/index.js:
```diff
@@ -0,0 +1,42 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import _ from 'lodash';
```
You are importing a very big library to use just one function (`map`).
Besides, map is a ES6 function that you can use without lodash.

Honestly, most of lodash's functions can be easily dropped in favour of ES6

sopasdigital replied:
Oh, that makes sense. I thought lodash was necessary for this case. I'll refactor this then :)

pdrbrnd commented on my-code/movie-search/src/Containers/MovieCatalogue/index.js:
```diff
@@ -0,0 +1,42 @@
+import _ from 'lodash';
+import { Link } from 'react-router-dom';
+
+import { fetchMovies } from '../../Actions';
```
You're importing this action creator but it's never used

sopasdigital replied:
Right, I wasn't able to take it out. Good catch!

pdrbrnd commented on my-code/movie-search/src/Containers/MovieCatalogue/index.js:
```diff
@@ -0,0 +1,42 @@
+import { fetchMovies } from '../../Actions';
+
+class MovieCatalogue extends Component {
+  renderCatalogue() {
```
There's no need for `renderCatalogue`.
This class could be a functional component with the mapping of the movies directly in the render:

```jsx
const MovieCatalogue = ({ moviesList }) => (
  <div className="MovieCatalogue">
    <ul>
      {moviesList.Search.map(movie => (
        <li key={movie.imdbID} className="Card">
        <Link to={`/movies/${movie.imdbID}`}>
          <div className="Card-Overlay">
            <div className="Card-Info">
              <h4 className="Card-Title">{movie.Title}</h4>
              <span className="Card-Year">{movie.Year}</span>
            </div>
          </div>
          <img className="Card-Poster" src={movie.Poster} alt={movie.Title} />
        </Link>
      </li>
      ))}
    </ul>
  </div>
);
```

sopasdigital replied:
I'll probably need you, Danilo or whoever to help me grasp the difference on this one. Thanks for pointing it out though!

pdrbrnd commented on my-code/movie-search/src/Containers/MovieDetails/index.js:
```diff
@@ -0,0 +1,66 @@
+    const { imdbID } = this.props.match.params;
+    this.props.fetchMovie(imdbID);
+
+    console.log(this.props);
```
Careful with the console.logs and commented code (assuming that this submission is like a PR to production)

sopasdigital replied:
Duly noted!

pdrbrnd commented on my-code/movie-search/src/Containers/MovieDetails/index.js:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  render() {
+    const movie = this.props.chosenMovie;
```
you could use ES6 destructuring: https://codeburst.io/es6-destructuring-the-complete-guide-7f842d08b98f?gi=d93f8df8d0a1

```jsx
const { chosenMovie: movie } = this.props;
```

sopasdigital replied:
This will be a must-read for me. I've only started understanding when I can destructure.

pdrbrnd commented on my-code/movie-search/src/Containers/MovieDetails/index.js:
```diff
@@ -0,0 +1,66 @@
+
+  render() {
+    const movie = this.props.chosenMovie;
+    const getValues = movie.Ratings&&movie.Ratings.map((review) => <span key={review.Source} className="Reviews">{review.Source} : {review.Value}</span> );
```
This is very long and hard to read. Consider using Prettier or something like it.
Also, using a linter would've helped you in general.

pdrbrnd commented on my-code/movie-search/src/Containers/MovieDetails/index.js:
```diff
@@ -0,0 +1,66 @@
+import React, { Component } from 'react';
```
As you're not dispatching several actions on your async calls, if we open a movie detail, go back and open another movie, we still see the old details for a second before it changes

sopasdigital replied:
Yes, I noticed this too. Your earlier recommendation on async catches this eventuality?

pdrbrnd commented on my-code/movie-search/src/Containers/SearchBar/index.js:
```diff
@@ -0,0 +1,34 @@
+
+  onInputChange(searchTerm) {
+    this.setState({ searchTerm });
+    this.props.fetchMovies(searchTerm);
```
We're making A LOT of requests here!
consider throttling or debouncing these requests

sopasdigital replied:
Right. I wasn't able to add the debounce on this one, my bad.

pdrbrnd commented on my-code/movie-search/src/Containers/SearchBar/index.js:
```diff
@@ -0,0 +1,34 @@
+        <input
+          placeholder="Search movies..."
+          value={this.state.searchTerm}
+          onChange={event => this.onInputChange(event.target.value)} />
```
You could just pass the function reference and get event.target.value there instead of creating a new anonymous function on each render

sopasdigital replied:
Noted!

pbrandone commented:
@sopasdigital Left some comments. @ruicsereno will also review your CSS

ruicsereno commented on my-code/movie-search/src/App.css:
```diff
@@ -0,0 +1,159 @@
+  font-weight: 700;
+}
+
+* {
```
I'd use this kind of generic rules at a Reset separate file.

ruicsereno commented on my-code/movie-search/src/App.css:
```diff
@@ -0,0 +1,159 @@
+  color: #7A8C99;
+}
+
+h1 {
```
Typography rules could be a separate file for better organisation

ruicsereno commented on my-code/movie-search/src/App.css:
```diff
@@ -0,0 +1,159 @@
+
+}
+
+.Card {
```
All Card rules could be in a separate file

ruicsereno commented on my-code/movie-search/src/App.css:
```diff
@@ -0,0 +1,159 @@
+.MovieCatalogue ul {
+  display: flex;
+  flex-wrap: wrap;
+  list-style: none;
```
List style could be solved at the Reset file. It would make CSS rules to the point and cleaner

ruicsereno commented on my-code/movie-search/src/App.css:
```diff
@@ -0,0 +1,159 @@
+  display: flex;
+  flex-wrap: wrap;
+  list-style: none;
+  padding: 0;
```
Same as list-style

ruicsereno commented on my-code/movie-search/src/App.css:
```diff
@@ -0,0 +1,159 @@
+  flex-wrap: wrap;
+}
+
+.MovieDetails-Information {
```
Each component, not only this ones and the ones I commented above could be separate, at different files. It would be easier to locate them and to work on it specifically.

sopasdigital replied:
I initially wanted to separate each of the css based on which components used them. I'll ask the team how they typically structure these files so I can apply that. Thanks for the feedback, Rui!

ruicsereno commented on my-code/movie-search/src/App.css:
```diff
@@ -0,0 +1,159 @@
+  width: 60%;
+}
+
+.MovieDetails-Information p {
```
This shouldn't be done. Typography should all be defined at a Typography file. Styling text like this, so specifically, is not error proof at all. Avoid such specific styling

sopasdigital committed:
Initial Commit

Created core framework using internal standards as a reference. Did a
preliminary check if linkages work by creating a simple layout that
outputs text. Will loop back to this and create proper components for
the pages.

sopasdigital committed:
Recoded the whole site

Recoded and refactored my initial code submission. It's still missing some functionalities
but I've implemented some new things based on the initial feedback from everyone. I still
need to understand better a lot of the concepts though.

sopasdigital committed:
Added icons for the Ratings

Added icons for the Movie Ratings and fixed some typos from the logo components

danilowoz commented on my-code/movie-search/src/actions/index.js:
```diff
@@ -0,0 +1,43 @@
+  FETCH_MOVIE_ERROR: "FETCH_MOVIE_ERROR"
+};
+
+const ROOT_URL = "http://www.omdbapi.com/?apikey=";
```
You can create an instance from Axios, then you don't need to set the URL every time:
https://github.com/axios/axios#axioscreateconfig

sopasdigital replied:
Oh nice! I'll take a look at this :)

danilowoz commented on my-code/movie-search/src/actions/index.js:
```diff
@@ -0,0 +1,43 @@
+};
+
+const ROOT_URL = "http://www.omdbapi.com/?apikey=";
+const API_KEY = "5766cd6e";
```
You should never send the private keys to Github, save it in a `.env` file
https://www.npmjs.com/package/dotenv

sopasdigital replied:
ah, noted!

danilowoz commented on my-code/movie-search/src/components/MovieCatalogue/index.js:
```diff
@@ -0,0 +1,35 @@
+  </Container>
+);
+
+function mapStateToProps({ movieCatalogue }) {
```
Prefer ES6 syntax:
```jsx
const mapStateToProps = ({ movieCatalogue }) => ({ movieCatalogue })
```

danilowoz commented on my-code/movie-search/src/components/UI/Provider/index.js:
```diff
@@ -0,0 +1,43 @@
+
+  body {
+    background-color: ${colors.greyscale.black};
+    font-family: 'Roboto', sans-serif;
```
Save the main font family in a variable

danilowoz commented on my-code/movie-search/src/components/UI/Provider/utils.js:
```diff
@@ -0,0 +1,30 @@
+import colors from "./colors";
```
Split the utils in different files

sopasdigital replied:
will do!

danilowoz commented on my-code/movie-search/src/components/UI/Provider/utils.js:
```diff
@@ -0,0 +1,30 @@
+import colors from "./colors";
+
+export const applyColor = props => {
```
I'm not a big fan of `swith case` :)
Maybe you can try something like this:

```js
export default ({color}) => {
  const variations = {
    yellow: colors.yellow,
    red: colors.red,
    ....
  }

  return colors[color] ? colors[color] : colors.greyscale.black
}
```

danilowoz commented on my-code/movie-search/src/containers/MovieDetails/index.js:
```diff
@@ -0,0 +1,113 @@
+    this.props.fetchMovie(imdbID);
+  }
+
+  convertToList(list, title) {
```
Prefer to use arrow function and you should always return something.
In this case, you can early return `null` whether `list` is `null` or `undefined`.

Take a look at this: https://forum.freecodecamp.org/t/the-return-early-pattern/19364

danilowoz commented on my-code/movie-search/src/containers/MovieDetails/index.js:
```diff
@@ -0,0 +1,113 @@
+    }
+  }
+
+  applyIcon(src) {
```
Same thing here, about the early return and `swith case`

danilowoz commented on my-code/movie-search/src/containers/MovieDetails/index.js:
```diff
@@ -0,0 +1,113 @@
+  }
+}
+
+function mapStateToProps(state) {
```
ES6 syntax please

danilowoz commented on my-code/movie-search/src/containers/SearchBar/index.js:
```diff
@@ -0,0 +1,49 @@
+    super(props);
+
+    this.state = { searchTerm: "" };
+    this.startFetch = debounce(this.startFetch, 300);
```
👍

danilowoz commented on my-code/movie-search/src/containers/SearchBar/index.js:
```diff
@@ -0,0 +1,49 @@
+    this.startFetch = debounce(this.startFetch, 300);
+  }
+
+  onInputChange = e => {
```
maybe: 
```js
onInputChange = ({target: { value: searchTerm }}) => {
    this.setState({ searchTerm }, this.startFetch);
}
```

danilowoz commented on my-code/movie-search/src/containers/SearchBar/index.js:
```diff
@@ -0,0 +1,49 @@
+  }
+}
+
+function mapStateToProps({ movieCatalogue }) {
```
ES6 syntax please

danilowoz commented on my-code/movie-search/src/index.js:
```diff
@@ -0,0 +1,34 @@
+  applyMiddleware(thunk)
+);
+
+// const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
```
remove the unnecessary comment

danilowoz commented on my-code/movie-search/src/index.js:
```diff
@@ -0,0 +1,34 @@
+import React from "react";
+import ReactDOM from "react-dom";
+import { Provider as Store } from "react-redux";
```
Save the redux things in a different file, then you will split the responsibilities of this file

danilowoz commented on my-code/movie-search/src/index.js:
```diff
@@ -0,0 +1,34 @@
+
+const store = createStore(
+  reducers,
+  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
```
Only add this line to dev mode, read more about environments mode in Node.js

sopasdigital replied:
Oh okay, I thought this was necessary for the actual app. Will note that :)

danilowoz commented on my-code/movie-search/src/reducers/reducerChosenMovie.js:
```diff
@@ -0,0 +1,13 @@
+import { types } from "../actions";
+
```
You can destructure the variable on the top of the function and please use ES6 syntax

```js
export default (state = {}, { type, payload }) => {
```

danilowoz commented on my-code/movie-search/src/reducers/reducerMovieCatalogue.js:
```diff
@@ -0,0 +1,13 @@
+import { types } from "../actions";
+
+export default function(state = {}, action) {
```
same thing here

danilowoz commented on my-code/movie-search/src/routes/index.js:
```diff
@@ -0,0 +1,14 @@
+import MovieSearch from "../views/MovieSearch";
+import MovieDetails from "../containers/MovieDetails";
+
+export default () => (
```
Remember to catch the page not found

sopasdigital replied:
I'll need your help and guidance for this one :)
